fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStore.java

```diff
@@ -1808,20 +1808,57 @@ private static boolean checkFirstKeyIsHeader(@Nonnull KeyValue firstKeyValue,
                                                  @Nonnull SubspaceProvider subspaceProvider,
                                                  @Nonnull Subspace subspace,
                                                  @Nonnull StoreExistenceCheck existenceCheck) {
-        if (TupleHelpers.equals(subspace.unpack(firstKeyValue.getKey()), Tuple.from(STORE_INFO_KEY))) {
+        final Tuple firstKey = subspace.unpack(firstKeyValue.getKey());
+        if (TupleHelpers.equals(firstKey, Tuple.from(STORE_INFO_KEY))) {
             return true;
         } else if (existenceCheck == StoreExistenceCheck.NONE) {
             LOGGER.warn(KeyValueLogMessage.of("Record store has no info but is not empty",
                     subspaceProvider.logKey(), subspaceProvider.toString(context),
-                    LogMessageKeys.KEY, subspace.unpack(firstKeyValue.getKey())));
+                    LogMessageKeys.KEY, firstKey));
             return false;
+        } else if (existenceCheck == StoreExistenceCheck.ERROR_IF_NO_INFO_AND_HAS_RECORDS_OR_INDEXES) {
+            final FDBRecordStoreKeyspace keyspace = determineRecordStoreKeyspace(firstKey, subspaceProvider, context);
+            // White list of acceptable key ranges for the first key. This may need to be updated as more keyspaces are added.
+            if (FDBRecordStoreKeyspace.INDEX_STATE_SPACE.equals(keyspace) || FDBRecordStoreKeyspace.INDEX_RANGE_SPACE.equals(keyspace)) {
+                LOGGER.warn(KeyValueLogMessage.of("Record store has no info or records but is not empty",
+                        subspaceProvider.logKey(), subspaceProvider.toString(context),
+                        LogMessageKeys.KEY, firstKey));
+                return false;
+            } else {
+                throw noInfoAndNotEmptyException(firstKey, subspaceProvider, context);
+            }
         } else {
-            throw new RecordStoreNoInfoAndNotEmptyException("Record store has no info but is not empty",
+            throw noInfoAndNotEmptyException(firstKey, subspaceProvider, context);
```

Comment:
`ERROR_IF_NO_INFO_AND_NOT_EMPTY` is not explicitly shown here. Maybe it would be more clear to change it to a switch on `existenceCheck `

Reply:
Oh, I misunderstood the comment. I thought this was saying that it wasn't clear in the returned error message whether it came from `ERROR_IF_NO_INFO_AND_NOT_EMPTY` or `ERROR_IF_NO_INFO_AND_HAS_RECORDS_OR_INDEXES`. I updated the `noInfoAndNotEmptyException` method to take a static method to try and address *that* concern. Oops.

In any case, I think the reason it is like this rather than a switch is that there are a couple of different existence checks that all want the same behavior here, e.g., `ERROR_IF_EXISTS` will still also throw an error if the store has no header but is not empty.

Reply:
Sounds good.